### PR TITLE
Parallel SGS, zone fixes

### PR DESCRIPTION
**Pull request by StannisMod**
Replaced Genetic's `Serial SGS` with `Parallel SGS`
- added `can_schedule_at_the_moment` method to all timelines
- updated `ZoneTimeline` with the ability to handle non-zero status change costs
- added `GeneralTimeline` to handle abstract objects in time
- added `LinkedList` data structure with iterators to handle remove-on-iteration operations

Parallel SGS is much more applicable to handle many constraints than Serial SGS because it depends on constraints checking method, not on find minimum satisfaction time.
In SAMPO we have renewable and non-renewable resources and zones, so it's true for us :)

### CODE REVIEW

File: sampo/scheduler/timeline/general_timeline.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+T = TypeVar('T')
+
+
+class GeneralTimeline(Generic[T]):
+    """
+    The representation of general-purpose timeline that supports some general subset of functions
+    """
+    def __init__(self):
+        # ScheduleEvent = time, idx, object
```
Comment (vanoha) [inline sampo/scheduler/timeline/general_timeline.py:16]: Commented code

File: sampo/scheduler/genetic/operators.py [hunk 2]
```diff
@@ -547,7 +548,7 @@ def mate(ind1: ChromosomeType, ind2: ChromosomeType, optimize_resources: bool, r
     """
     child1, child2 = mate_scheduling_order(ind1, ind2, rand, copy=True)
     child1, child2 = mate_resources(child1, child2, rand, optimize_resources, copy=False)
-    child1, child2 = mate_for_zones(child1, child2, rand, copy=False)
+    # child1, child2 = mate_for_zones(child1, child2, rand, copy=False)
```
Comment (vanoha) [inline sampo/scheduler/genetic/operators.py:551]: Write TODO

File: sampo/scheduler/genetic/operators.py [hunk 3]
```diff
@@ -573,7 +574,7 @@ def mutate(ind: ChromosomeType, resources_border: np.ndarray, parents: dict[int,
     """
     mutant = mutate_scheduling_order(ind, order_mutpb, rand, parents, children)
     mutant = mutate_resources(mutant, res_mutpb, rand, resources_border)
-    mutant = mutate_for_zones(mutant, zone_mutpb, rand, statuses_available)
+    # mutant = mutate_for_zones(mutant, statuses_available, zone_mutpb, rand)
```
Comment (vanoha) [inline sampo/scheduler/genetic/operators.py:577]: Write TODO

File: sampo/scheduler/timeline/base.py [hunk 4]
```diff
@@ -67,15 +67,21 @@ def find_min_start_time_with_additional(self,
             -> tuple[Time, Time, dict[GraphNode, tuple[Time, Time]]]:
         ...
 
+    @abstractmethod
+    def can_schedule_at_the_moment(self,
```
Comment (vanoha) [inline sampo/scheduler/timeline/base.py:71]: Add docs for method

**Review by vanoha — COMMENTED** [review]
Request changes

File: sampo/scheduler/timeline/general_timeline.py (hunk 1 above)
Comment (StannisMod) [reply sampo/scheduler/timeline/general_timeline.py:16]: It's not code, it's comment :)

**Review by vanoha — APPROVED** [review]
Approved

**Pull request merged**